File: cloud/Server/API/MetricController.cs
// Copyright © Spatial Corporation. All rights reserved.

using Spatial.Cloud.Data.Scopes;
using Spatial.Identity.Authorization;
using Spatial.Persistence;

namespace Spatial.Cloud.API;

/// <summary>
/// A <see cref="Controller"/> for metric data.
/// </summary>
[Path("metrics")]
public class MetricController : Controller
{
    /// <summary>
    /// Reads data points for the specified metric.
    /// </summary>
    /// <param name="name">The metric's name.</param>
    /// <param name="from">Optional start timestamp (UTC).</param>
    /// <param name="to">Optional end timestamp (UTC).</param>
    /// <param name="limit">Optional maximum number of records to return.</param>
    /// <param name="resolution">Optional search resolution.</param>
    /// <returns>A list of matching <see cref="Metric"/> records.</returns>
    [GET]
    [Path("{name}")]
    [Authorize(Scope.Metrics.Read)]
    public Task<List<Metric>> ReadAsync(
        string name,
        [Query] DateTime? from = null,
        [Query] DateTime? to = null,
        [Query] int? limit = null,
        [Query] string? resolution = null)
    {
        return Metric.ReadAsync(name, from, to, limit, resolution);
    }
}

File: interface/client/controllers/MetricController.ts
// Copyright © Spatial Corporation. All rights reserved.

import { Metric } from "@sptlco/data";
import { Controller } from "..";

export class MetricController extends Controller {
  /**
   * Read data points for a given metric.
   * @param name The metric's name.
   * @param from Optional start date (UTC).
   * @param to Optional end date (UTC).
   * @param limit Optional maximum number of records.
   * @returns Matching data points.
   */
  public read = (name: string, from?: Date, to?: Date, limit?: number, resolution?: string) => {
    const params = new URLSearchParams();

    if (from) {
      params.append("from", from.toISOString());
    }

    if (to) {
      params.append("to", to.toISOString());
    }

    if (limit != null) {
      params.append("limit", limit.toString());
    }

    if (resolution) {
      params.append("resolution", resolution);
    }

    const query = params.toString();

    return this.get<Metric[]>(query ? `metrics/${name}?${query}` : `metrics/${name}`);
  };
}
